pkg/cloudlogging/puller_test.go

```diff
@@ -98,13 +98,117 @@ func TestPull(t *testing.T) {
 				t.Errorf("Process(%+v) got result diff (-want, +got): %v", tc.name, diff)
 			}
 
-			if diff := cmp.Diff(tc.wantReq, tc.server.req, protocmp.Transform()); diff != "" {
+			if diff := cmp.Diff(tc.wantReq, tc.server.listReq, protocmp.Transform()); diff != "" {
 				t.Errorf("Process(%+v) got request diff (-want, +got): %v", tc.name, diff)
 			}
 		})
 	}
 }
 
+func TestStreamPull(t *testing.T) {
+	t.Parallel()
+
+	cases := []struct {
+		name          string
+		server        *fakeServer
+		filter        string
+		wantReq       *loggingpb.TailLogEntriesRequest
+		wantResult    []*loggingpb.LogEntry
+		wantNum       int
+		wantErrSubstr string
+	}{
+		{
+			name:       "success",
+			filter:     "test-filter",
+			server:     &fakeServer{logNumCap: 10000},
+			wantResult: []*loggingpb.LogEntry{{LogName: "test-0"}, {LogName: "test-1"}},
+			wantReq: &loggingpb.TailLogEntriesRequest{
+				ResourceNames: []string{testResource},
+				Filter:        "test-filter",
+			},
+			wantNum: 2,
+		},
+		{
+			name:   "failed_to_pull",
+			filter: "test-filter",
+			server: &fakeServer{
+				injectedErr: fmt.Errorf("injected error"),
+			},
+			wantReq: &loggingpb.TailLogEntriesRequest{
+				ResourceNames: []string{testResource},
+				Filter:        "test-filter",
+			},
+			wantErrSubstr: "injected error",
+			wantNum:       0,
+		},
+	}
+
+	for _, tc := range cases {
+		tc := tc
+
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			ch := make(chan *loggingpb.LogEntry, 2)
+			var gotLogs []*loggingpb.LogEntry
+			done := make(chan struct{}, 1)
+			t.Cleanup(func() {
+				close(done)
+			})
+
+			ctx, cancel := context.WithCancel(context.Background())
+			// create a go routine to receive logs sent from
+			// streamPull,  break after receive enough logs.
+			go func() {
+				for l := range ch {
+					gotLogs = append(gotLogs, l)
+					if len(gotLogs) == tc.wantNum {
+						done <- struct{}{}
+						break
+					}
+				}
```

Comment:
I think I found an easier way to do this. Can you try it out?

```go
ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)

go func() {
  for l := range ch {
    gotLogs = append(gotLogs, l)
    if len(gotLogs) == tc.wantNum {
      cancel() // got enough logs, we can stop now
      break
    }
  }
}

// set up fake client

go func() {
  defer close(ch)
  gotPullErr = p.StreamPull(ctx, tc.filter, ch)
}

<-ctx.Done() // Either we timed out or we got enough logs and explicitly cancelled it

// Do the checks
```

Reply:
Tested and updated.